### ISSUE TITLE
use cmake instead of waf ?

### ISSUE THREAD

**Issue by mrVanDalo** [jackaudio/jack2#236]
I'm not an expert in build tools, but always when I'm seeing _waf_ I lose joy in joining a project.

This is mainly because awesome IDEs like CLion or VisualStudio come with a very good support for cmake, but no support for _waf_ at all.

I would be delighted to join this project, but I would only join if cmake is the build-tool.
I don't think I'm alone with that opinion :smile:

**Comment by mseeber** [comment]
If that is so important to you, are you sure you want to work on JACK and not a generic other project that uses CMake instead?

You could write a build system for jack2 with CMake though, maybe it will be better than the current waf approach.
Asking the current developers though, is really a stretch since this involves a lot of work (re engineering, implementation, porting, testing, bugfixing) and risk.

Anyway, @sletz @lilrc probably are more qualified to judge that.

**Comment by lilrc** [comment]
I agree with @mseeber.

What's more, I don't see why this is necessary. In fact changing the build system just creates more work. You would have to firstly do all the work @mseeber points out and secondly all downstream maintainers would have to update their build scripts.

I would argue in the line of "if it aint broke, don't fix it" and that some IDE having support for CMake but not waf i don't qualify as the current build system being broke. What kind of support is it that these IDEs supply, anyway?

**Comment by sletz** [comment]
Waf stuff was code by Nedko initially. Please do not start « yet another build tool war »

**Comment by nedko** [comment]
Ingolf Wagner <notifications@github.com> writes:

> I'm not an expert in build tools, but always when I'm seeing _waf_ I
> lose joy in joining a project.
>
> This is mainly because awesome IDEs like CLion or VisualStudio come
> with a very good support for cmake, but no support for _waf_ at all.
>
> I would be delighted to join this project, but I would only join if
> cmake is the build-tool.
> I don't think I'm alone with that opinion :smile: 

Markus Seeber <notifications@github.com> writes:

> If that is so important to you, are you sure you want to work on JACK
> and not a generic other project that uses CMake instead?
>
> You could write a build system for jack2 with CMake though, maybe it
> will be better than the current waf approach.
> Asking the current developers though, is really a stretch since this
> involves a lot of work (re engineering, implementation, porting,
> testing, bugfixing) and risk.
>
> Anyway, @sletz @lilrc probably are more qualified to judge that.

Karl Lindén <notifications@github.com> writes:

> I agree with @mseeber.
>
> What's more, I don't see why this is necessary. In fact changing the
> build system just creates more work. You would have to firstly do all
> the work @mseeber points out and secondly all downstream maintainers
> would have to update their build scripts.
>
> I would argue in the line of "if it aint broke, don't fix it" and that
> some IDE having support for CMake but not waf i don't qualify as the
> current build system being broke. What kind of support is it that
> these IDEs supply, anyway?

Ingolf, you can write the cmake files, if you beleive in cmake
superiourity so much. It can be probably even included upstream, given
that it does not need changes in source files (a must for any build
system I would call sane for use). So I agree with Markus.

Also, as Karl already pointed out, downstream distro maintainers will
have to be conviced to do the work. And that is the really challenging
part.

IIRC, waf is not actually used for building on the microsoft os and on
the apple OS. I think waf was usable at some point on the apple os. But
there is XCode project. For the windows os, IIRC there is a codeblocks
project. So I guess adding cmake (if it is sane) wouldn't make the jack2
build system arena much worse :)

So in short, you better show your work or not smile as a troll using
bandwagon :P

**Comment by mrVanDalo** [comment]
First things first : « yet another build tool war » 

I can understand, you don't want to switch. I just wanted to ask before I start creating a cmake system and realize in the pull-request that you guys don't want it (which is obviously the case).

I think the best way you keep `waf` and I find myself another project to support.

Here are some answers to the questions : 

> What kind of support is it that these IDEs supply, anyway?
Your IDE can help you with marco problems, refactorings, debugging and a lot of things more, but it needs a cmake file for that.

> this involves a lot of work
For sure, this is why I'm asking before starting a pull-request.

> downstream maintainers would have to update their build scripts
That's not a big problem I think because building via cmake is pretty much settled in across all distributions.

> given that it does not need changes in source files (a must for any build system I would call sane for use).
I agree.

> waf is not actually used for building on the microsoft os and on the apple OS
That's something cmake is good in.

> smile as a troll using bandwagon
äähm I'm ignoring this comment from now on.

**Comment by jcelerier** [comment]
@mrVanDalo  did you ever happen to write such CMake scripts ? It would be fairly useful to allow building with recent MSVC versions without having to install a whole code::blocks IDE.

**Comment by mrVanDalo** [comment]
@jcelerier No sorry Did not.

**Comment by HiImJulien** [comment]
Sorry for being a necromancer on this issue.

@jcelerier The *supposed great thing* about CMake is its ability to generate project files for any IDE you can think of. Additionally it generates build scripts, such as Makefile or Ninja build scripts. 

To address your issue: With CMake you can generate project files for Visual Studio and use MSVC to compile your project. For the unusual case, that you're using command line tools without Visual Studio, you can use msbuild (which in turn uses VS projects to build). 

It is important to state, that CMake has, IMHO, the worst dsl one would possibly able to come up. It's just been there at the right time, leading to the white spread adoption we're seeing now.
